### PR TITLE
Fm classifier

### PR DESCRIPTION
**Pull request by Hydrotoast**
Some major changes:
1. Added an abstract `FMPredictor` where `FMRegressor` and `FMClassifier` are subtypes.
2. `FMRegressor` prediction and loss functions live in `fm_regressor.jl`
3. `FMClassifier` prediction and loss functions live in `fm_classifier.jl`
4. Common code for `FMRegressor` and `FMClassifier` lives in `fm.jl`
5. To train a classifier instead of a regressor, use `fmTrain(X, y, task = :classification)`

### CODE REVIEW

**Comment by Hydrotoast** [comment]
Related to #2

File: src/fm_classifier.jl [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+           sum_sqr[f] += d*d
+       end
+       result += 0.5 * (fSum[f]*fSum[f] - sum_sqr[f])
+   end
+    # squash result between [0, 1]
+    1.0 / (1.0 + exp(-result))
+end
+
+fmLoss(::Type{FMClassifier}, yhat::FMFloat, y::FMFloat) = (1 - sign(yhat * y)) / 2
+fmLossGradient(::Type{FMClassifier}, yhat::FMFloat, y::FMFloat) = -y * (1.0 - 1.0 / (1.0 + exp(-y * yhat)))
```
Comment (Hydrotoast) [inline src/fm_classifier.jl]: Should probably note that this is the gradient with respect to the prediction (not the gradient with respect to the parameters of the prediction).

File: src/fm.jl [hunk 2]
```diff
@@ -1,111 +1,19 @@
...
-    V = randn(num_factor, attributesNumber) .* initStd
-
-    #new model
-    FMModel(k0, k1, w0, w1, V, reg0, regw, regv, initMean, initStd, attributesNumber, num_factor, targetMin, targetMax )
-end
-
 function fmTrain(
     X::FMMatrix,
-    y::Array{FMFloat},
+    y::Array{FMFloat};
```
Comment (Hydrotoast) [inline src/fm.jl]: Everything else is keyword arguments so that position does not matter

File: src/fm_classifier.jl (hunk 1 above)
Comment (btwardow) [inline src/fm_classifier.jl]: maybe change signature of the method. However, for me it's OK.

**Comment by btwardow** [comment]
Ok. I haven't been programming in Julia for a few months so looking at the code was nice difference :) One is noticeable - lots of the same code repetition between body of FMClassifier and FMRegressor. Maybe there is a decent way to avoid this?

**Comment by Hydrotoast** [comment]
Agreed. There should be less code duplication. What's odd about Julia is that fields cannot be inherited (abstract types cannot have fields and subtypes cannot inherit from each other), so the option I'm leaning towards is simple encapsulation.

**Comment by Hydrotoast** [comment]
So I ended up refactoring a lot of the code and I think I have an extendable API. The best way to see this API is in the tests. Here is an example of building a classifier with Gaussian distributed parameters for the model vectors and SGD with an `alpha = 0.3`. Defaults are set where parameters are not specified e.g. the model has a default `num_factors = 8`.

I'll document this behavior in the wiki. Ideally, this approach should make it easy to add more _learning methods_ by adding them to the `Methods` submodule, _tasks_ by adding them to the `Tasks` submodule (e.g. ranking is also a common task using the BPR optimization criteria)

Here is some sample usage from the tests:

``` julia
using FactorizationMachines

# Matrix constructed as the outer product of two vectors: [-1.; 1.] and [1. 1.]
num_factors = 4
X = [
1. 1. 0. 0.;
0. 0. 1. 1.;
1. 0. 1. 0.;
0. 1. 0. 1.
]
y = [0., 1., 0., 1.]

info("Decomposing Rank 1 Matrix")
fm = train(sparse(X), y;
    method = Methods.sgd(alpha = 0.3),
    model_params = Models.gauss(k0 = false, k1 = false),
    task_params = Tasks.classification())
@test fm.model.w0 == 0
@test fm.model.w == zeros(num_factors)
@test evaluate(Evaluators.z1(), fm, sparse(X), y) == 0.0
```

The training task now returns an `FMPredictor`:

``` julia
type FMPredictor{T<:PredictorTask}
    task::T
    model::FMModel
end
```

which knows how to `predict(predictor, x)` for the appropriate task.

Let me know what you think of this.

**Comment by btwardow** [comment]
Ok. I will try to review and mare this today.

File: src/common.jl [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+export sqerr,
+       sqerr_deriv,
+       nlogsig,
+       nlogsig_deriv,
+       heaviside,
+       sigmoid
+
+"""Squared error"""
+sqerr(p::Number, y::Number)       = (p - y)^2
+sqerr_deriv(p::Number, y::Number) = (p - y)
```
Comment (btwardow) [inline src/common.jl:11]: To be the math purist - missing 2*

File: test/runtests.jl [hunk 4]
```diff
@@ -1,6 +1,10 @@
 using FactorizationMachines
 using Base.Test
 
-include("test_fm_readlibsvm.jl")
+include("test_read_libsvm.jl")
 include("test_fm.jl")
```
Comment (btwardow) [inline test/runtests.jl:5]: There is still `fmTrain`. Test is successful because the exception is expected. However, the different one is given ;-)

**Comment by btwardow** [comment]
@Hydrotoast great job! Huge PR but amazing refactoring. I look though the code and I think only single test in `test_fm.jl` needs to be fix and PR will be ready to merge.

**Comment by btwardow** [comment]
As this is a minor issue I'm merging it.

**Comment by Hydrotoast** [comment]
Cool, thanks for this. I will send a short PR with some quick fixes e.g. README update and removing that file for now. Should probably TODO for more tests.

File: src/common.jl (hunk 3 above)
Comment (Hydrotoast) [inline src/common.jl:11]: This will technically affect the learning rate, so I will add this in so that hyperparameters are comparable across implementations (assuming that other implementations are math purists).
